### PR TITLE
Offline downloads happen on a separate thread now.

### PR DESCRIPTION
**Pull request by cammace**
This Introduces an OfflineDownloadThread class which handles the offline region downloading and limits the number of callbacks that occur on the main thread (mainly status updates). I'd like to take this a step further in a different branch by allowing multiple regions to be downloaded at the same time.

### CODE REVIEW

**Comment by cammace** [comment]
This plugin is now blocked by https://github.com/mapbox/mapbox-gl-native/issues/11461. While the issue is being resolved, I'll be continuing work on cleaning this PR up and documenting the plugin in preparation for release.

**Comment by cammace** [comment]
I'm ready for another review on this PR when you get a chance @tobrun

File: app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/offline/OfflineRegionDetailActivity.java [hunk 1]
```diff
@@ -188,36 +188,36 @@ public void onFabClick(View view) {
   }
 
   @Override
-  public void onCreate(OfflineDownloadOptions offlineDownload) {
-    Timber.e("OfflineDownloadOptions created %s", offlineDownload.hashCode());
+  public void onCreate(DownloadOptions offlineDownload) {
```
Comment (langsmith) [inline app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/offline/OfflineRegionDetailActivity.java]: Maybe change the parameter name to `downloadOptions` or something without the word offline in it if the class name is no longer going to have the word `Offline` in it...

File: plugin-offline/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -6,5 +6,7 @@
     <activity
       android:name="com.mapbox.mapboxsdk.plugins.offline.ui.OfflineActivity"
       android:configChanges="keyboard|keyboardHidden|orientation|screenLayout|uiMode|screenSize|smallestScreenSize"/>
+
```
Comment (langsmith) [inline plugin-offline/src/main/AndroidManifest.xml]: nit line

File: plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineRegionSelector.java [hunk 3]
```diff
@@ -15,6 +17,32 @@ private OfflineRegionSelector() {
     // No Instances
   }
 
+  public static DownloadOptions getDownloadOptions(final Intent data, byte[] metadata) {
```
Comment (langsmith) [inline plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineRegionSelector.java]: Add a ticket to this repo documenting that explanatory javadocs should be written for this class and its methods?

**Comment by tobrun** [comment]
A couple of more things that needs to be fixed before we can merge this PR:

##### Cancel 

Clicking cancel does stop the download but doesn't dismiss the notification

ezgif com-video-to-gif 56

##### MapSnapshot

I create two downloads with different styles but the same snapshot is used

ezgif com-gif-maker 2

##### Detail activity

The boundable detail activity is broken

ezgif com-video-to-gif 57

##### Multiple threads

Since there still some more work to be done before we can merge this let's move to a single thread model instead of creating a new thread for each download.

File: plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineRegionSelector.java (hunk 3 above)
Comment (cammace) [reply plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineRegionSelector.java]: https://github.com/mapbox/mapbox-plugins-android/issues/385 addresses this.

**Comment by cammace** [comment]
Closing infavor of #407

**Pull request closed without merging**